lms/djangoapps/shoppingcart/views.py

```diff
@@ -348,27 +349,41 @@ def register_code_redemption(request, registration_code):
     elif request.method == "POST":
         reg_code_is_valid, reg_code_already_redeemed, course_registration = get_reg_code_validity(registration_code,
                                                                                                   request, limiter)
-
         course = get_course_by_id(getattr(course_registration, 'course_id'), depth=0)
+        context = {
+            'reg_code': registration_code,
+            'site_name': site_name,
+            'course': course,
+            'reg_code_is_valid': reg_code_is_valid,
+            'reg_code_already_redeemed': reg_code_already_redeemed,
+        }
         if reg_code_is_valid and not reg_code_already_redeemed:
             #now redeem the reg code.
-            RegistrationCodeRedemption.create_invoice_generated_registration_redemption(course_registration, request.user)
-            CourseEnrollment.enroll(request.user, course.id)
-            context = {
-                'redemption_success': True,
-                'reg_code': registration_code,
-                'site_name': site_name,
-                'course': course,
-            }
+            redemption = RegistrationCodeRedemption.create_invoice_generated_registration_redemption(course_registration, request.user)
+            try:
+                kwargs = {}
+                if course_registration.mode_slug is not None:
+                    if CourseMode.mode_for_course(course.id, course_registration.mode_slug):
+                        kwargs['mode'] = course_registration.mode_slug
+                    else :
+                        context['redeem_code_error'] = True
+                        context['redemption_success'] = False
+                        return render_to_response(template_to_render, context)
```

Comment:
OK, it took me awhile to figure out this control flow. Since we have a bunch of exception handling below that renders an error message in the view, can I suggest that you raise some typed exception here and handle it below keeping the implementation patters consistent?